### ISSUE TITLE
`Expression type contains "Any"` false positive when casting tuple

### ISSUE THREAD

**Issue by DetachHead** [python/mypy#16170]
```py
from typing import cast

cast(tuple[int, int], tuple(value for value in [1, 2]))
```
```
main.py:3: error: Expression type contains "Any" (has type "Generator[Any, None, None]")  [misc]
```
[playground](https://mypy-play.net/?mypy=latest&python=3.11&flags=show-error-context%2Cshow-error-codes%2Cstrict%2Ccheck-untyped-defs%2Cdisallow-any-decorated%2Cdisallow-any-expr%2Cdisallow-any-explicit%2Cdisallow-any-generics%2Cdisallow-any-unimported%2Cdisallow-incomplete-defs%2Cdisallow-subclassing-any%2Cdisallow-untyped-calls%2Cdisallow-untyped-decorators%2Cdisallow-untyped-defs%2Cno-implicit-reexport%2Clocal-partial-types%2Cstrict-equality%2Cwarn-incomplete-stub%2Cwarn-redundant-casts%2Cwarn-return-any%2Cwarn-unreachable%2Cwarn-unused-configs%2Cwarn-unused-ignores%2Cno-implicit-optional&gist=9d04feffcc941832f8896553175a5e3d)

**Comment by KotlinIsland** [comment]
minified
```py
from typing import cast, _T as T, Generic

class A(Generic[T]):
    def __init__(self, t: T): ...

cast(object, A(x for x in [1]))
```
